src/main/java/com/salesforce/dynamodbv2/mt/mappers/sharedtable/impl/RandomPartitioningConditionMapper.java

```diff
@@ -30,143 +38,272 @@
+            String virtualField = fieldMapping.getSource().getName();
+            Set<String> fieldPlaceholders = getPlaceholdersForVirtualField(virtualField, request); // {#field1}
+
+            for (String fieldPlaceholder : fieldPlaceholders) {
+                // TODO support one virtual field placeholder having multiple comparison clauses in an expression
+                Optional<String> valuePlaceholderOpt =
+                    findVirtualValuePlaceholder(expression, fieldPlaceholder); // "Optional[:value]"
+                if (valuePlaceholderOpt.isPresent()) {
+                    String valuePlaceholder = valuePlaceholderOpt.get();
+                    AttributeValue virtualValue = request.getExpressionAttributeValues()
+                        .get(valuePlaceholder); // {S: hkValue,}
+                    AttributeValue physicalValue = fieldMapping.isContextAware()
+                        ? fieldMapper.apply(fieldMapping, virtualValue) // {S: ctx.virtualTable.hkValue,}
+                        : virtualValue;
+                    request.putExpressionAttributeValue(valuePlaceholder, physicalValue);
+                }
+                request.putExpressionAttributeName(fieldPlaceholder, fieldMapping.getTarget().getName());
+            }
+        }
+    }
+
+    @VisibleForTesting
+    String mapFieldInUpdateExpression(String expression, // "#field1 = :value"
+                                      RequestWrapper request,
+                                      String virtualField, // "virtualHk"
+                                      List<FieldMapping> fieldMappings) {
+        if (expression != null) {
+            Set<String> virtualFieldPlaceholders = getPlaceholdersForVirtualField(virtualField, request); // {#field1};
+            for (String virtualFieldPlaceholder : virtualFieldPlaceholders) {
+                // each placeholder should appear at most once in update expression, assuming we support only SET
                 Optional<String> virtualValuePlaceholderOpt =
-                    findVirtualValuePlaceholder(primaryExpression, filterExpression, keyFieldName.get()); // ":value"
+                    findVirtualValuePlaceholder(expression, virtualFieldPlaceholder); // "Optional[:value]"
                 if (virtualValuePlaceholderOpt.isPresent()) {
                     String virtualValuePlaceholder = virtualValuePlaceholderOpt.get();
-                    AttributeValue virtualAttr =
+                    AttributeValue virtualValue =
                         request.getExpressionAttributeValues().get(virtualValuePlaceholder); // {S: hkValue,}
-                    AttributeValue physicalAttr =
-                        fieldMapping.isContextAware()
-                            ? fieldMapper.apply(fieldMapping, virtualAttr) // {S: ctx.virtualTable.hkValue,}
-                            : virtualAttr;
-                    request.putExpressionAttributeValue(virtualValuePlaceholder, physicalAttr);
+
```

Comment:
Below is how we fan out from one field assignment in the virtual SET expression to possibly multiple assignments in the physical SET expression